base/src/main/java/org/aya/core/visitor/TermOps.java

```diff
@@ -132,4 +143,105 @@ public Term post(Term term) {
       };
     }
   }
+
+  record Normalizer(@NotNull @Override TermView view, TyckState state) implements TermOps {
+    static private @NotNull Subst buildSubst(
+      @NotNull SeqLike<Term.@NotNull Param> params,
+      @NotNull SeqLike<@NotNull Arg<@NotNull Term>> args
+    ) {
+      var subst = new Subst(MutableMap.create());
+      params.view().zip(args).forEach(t -> subst.add(t._1.ref(), t._2.term()));
+      return subst;
+    }
+
+    private @Nullable WithPos<Term> unfoldClauses(
+      boolean orderIndependent, SeqLike<Arg<Term>> args,
+      @NotNull ImmutableSeq<Matching> clauses
+    ) {
+      return unfoldClauses(orderIndependent, args, new Subst(MutableMap.create()), clauses);
+    }
+
+    private @Nullable WithPos<Term> unfoldClauses(
+      boolean orderIndependent, SeqLike<Arg<Term>> args,
+      @NotNull Subst subst, @NotNull ImmutableSeq<Matching> clauses
+    ) {
+      for (var match : clauses) {
+        var result = PatMatcher.tryBuildSubstArgs(null, match.patterns(), args);
+        if (result.isOk()) {
+          subst.add(result.get());
+          var body = match.body().view().rename().subst(subst).normal(state).commit();
+          return new WithPos<>(match.sourcePos(), body);
+        } else if (!orderIndependent && result.getErr())
+          return null;
+      }
+      return null;
+    }
+
+    @Override public Term post(Term term) {
+      return switch (view.post(term)) {
+        case ElimTerm.App app -> {
+          var fn = app.of();
+          if (fn instanceof IntroTerm.Lambda lambda)
+            yield CallTerm.make(lambda, app.arg());
+          else yield app;
+        }
+        case ElimTerm.Proj proj -> {
+          var tup = proj.of();
+          var ix = proj.ix();
+          if (tup instanceof IntroTerm.Tuple tuple) {
+            assert tuple.items().sizeGreaterThanOrEquals(ix) && ix > 0
+              : proj.toDoc(DistillerOptions.debug()).debugRender();
+            yield tuple.items().get(ix - 1);
+          } else yield proj;
+        }
+        case CallTerm.Con con -> {
+          var def = con.ref().core;
+          if (def == null) yield con;
+          var unfolded = unfoldClauses(true, con.conArgs(), def.clauses);
+          yield unfolded != null ? unfolded.data() : con;
+        }
+        case CallTerm.Fn fn -> {
+          var def = fn.ref().core;
+          if (def == null) yield fn;
+          if (def.modifiers.contains(Modifier.Opaque)) yield fn;
+          yield def.body.fold(
+            lamBody -> lamBody.view().subst(buildSubst(def.telescope(), fn.args())).rename().normal(state).commit(),
```

Comment:
Do we really need to `rename` here, given that our `subst` has capture-avoiding?

Reply:
(note that my comment also applies to other `rename` calls)

Reply:
Ahh, the renaming is being done at substitution now, right?

Reply:
Yaas